base/reducedim.jl

```diff
@@ -27,271 +37,323 @@ function reduced_indices0(axs::Indices{N}, region) where N
+# loops.  The trouble is that the naive double for loop is leaving 100x+ performance on the
+# table when it doesn't iterate in a cache-friendly manner.  So you might indeed want to update
+# a single value in `R` multiple times — and possibly widen as you do so.
+#
+# It's **_so very tempting_** to do the standard Julian idiom of:
+#     mapreduce(f, op, A, dims) = mapreduce!(f, op, init(f, op, A, dims), A)
+# But that simply doesn't work because there's not _a single way_ to initialize the beginning 
+# of a reduction.  There are SIX.  But then we add support for pre-allocated result vectors and
+# you even _moreso_ want to be able to just write, e.g.,
+#     sum!(r, A) = mapreduce!(identity, +, first_step!(identity, +, r, A), A)
+# But at that point you've _already done the first step_, so you need to be able to _continue_
+# the reduction from an arbitrary point in the iteration.
+_mapreducedim_impl(f::Type{F}, op, init, A::AbstractArrayOrBroadcasted, raxs, out) where {F} =
+    _mapreducedim_impl(x->F(x), op, init, A::AbstractArrayOrBroadcasted, raxs, out)
+function _mapreducedim_impl(f, op, init, A::AbstractArrayOrBroadcasted, raxs, out)
+    # @show f, op, init, A, raxs, out
+    outer_inds = CartesianIndices(raxs)
+    a_inds = CartesianIndices(axes(A))
+    Aaxs = axes(A)
+    @assert length(Aaxs) == length(raxs)
+    # The "outer dimension" of the reduction are the preserved dimensions
+    is_outer_dim = map((>(1))∘length, raxs)
+    n = prod(map((b, ax)->ifelse(b, 1, length(ax)), is_outer_dim, Aaxs))
+
+    # We have special support for returning empty arrays
+    # isempty(outer_inds) && return collect_allocator(_mapreduce_similar_curried(A),
+    #     (_mapreduce_naive_inner_loop(f, op, init, A, is_outer_dim, outer_ind, a_inds) for outer_ind in outer_inds))
+    (n == 0 || isempty(A)) && return _mapreduce_empty_array(f, op, init, A, raxs, out)
+    n == 1 && return _mapreduce_one_array(f, op, init, A, raxs, out)
+
+    _mapreduce_might_widen(f, op, A, out) && return collect_allocator(_mapreduce_similar_curried(A),
+        (_mapreduce_naive_inner_loop(f, op, init, A, is_outer_dim, outer_ind, a_inds) for outer_ind in outer_inds))
+
+    # Create the result vector with the first step of the reduction, using the passed output if given
+    # note that this is using a (potentially) bad cache ordering, so we don't want to continue like this
+    i1 = first(outer_inds)
+    inner1 = mergeindices(is_outer_dim, i1, a_inds)
+    i1, i2 = inner1
+    @inbounds a1, a2 = A[i1], A[i2]
+    v = _mapreduce_start(f, op, init, a1, a2)
+    R = isnothing(out) ? mapreduce_similar(A, typeof(v), raxs) : out
+    @inbounds R[i1] = v
+    for i in Iterators.drop(outer_inds, 1)
+        innerj = mergeindices(is_outer_dim, i, a_inds)
+        j1, j2 = innerj
+        @inbounds c1, c2 = A[j1], A[j2]
+        @inbounds R[i] = _mapreduce_start(f, op, init, c1, c2)
+    end
```

Comment:
This loop might need manual inner loop expansion, as it blocks possible vectorization if there's no reduction on dim1. Another solution is adding `@simd` support to `drop(::CartesianIndices)` https://github.com/N5N3/julia/blob/fd1b3a19770d4948451c2d32cf894ea1b9536038/base/multidimensional.jl#L703-L716